File: src/libcadet/model/binding/AntiLangmuirBinding.cpp
// =============================================================================
//  CADET
//  
//  Copyright © The CADET Authors
//            Please see the CONTRIBUTORS.md file.
//  
//  All rights reserved. This program and the accompanying materials
//  are made available under the terms of the GNU Public License v3.0 (or, at
//  your option, any later version) which accompanies this distribution, and
//  is available at http://www.gnu.org/licenses/gpl.html
// =============================================================================

#include "model/binding/BindingModelBase.hpp"
#include "model/ExternalFunctionSupport.hpp"
#include "model/ModelUtils.hpp"
#include "cadet/Exceptions.hpp"
#include "model/Parameters.hpp"
#include "LocalVector.hpp"
#include "SimulationTypes.hpp"

#include <vector>
#include <unordered_map>
#include <functional>

/*<codegen>
{
	"name": "AntiLangmuirParamHandler",
	"externalName": "ExtAntiLangmuirParamHandler",
	"parameters":
		[
			{ "type": "ScalarComponentDependentParameter", "varName": "kA", "confName": "MCAL_KA"},
			{ "type": "ScalarComponentDependentParameter", "varName": "kD", "confName": "MCAL_KD"},
			{ "type": "ScalarComponentDependentParameter", "varName": "qMax", "confName": "MCAL_QMAX"},
			{ "type": "ScalarComponentDependentParameter", "varName": "antiLangmuir", "confName": "MCAL_ANTILANGMUIR"}
		]
}
</codegen>*/

/* Parameter description
 ------------------------
 kA = Adsorption rate
 kD = Desorption rate
 qMax = Capacity
 antiLangmuir = Anti-Langmuir factor
*/

namespace cadet
{

namespace model
{

inline const char* AntiLangmuirParamHandler::identifier() CADET_NOEXCEPT { return "MULTI_COMPONENT_ANTILANGMUIR"; }

inline bool AntiLangmuirParamHandler::validateConfig(unsigned int nComp, unsigned int const* nBoundStates)
{
	if ((_kA.size() != _kD.size()) || (_kA.size() != _qMax.size()) || (_kA.size() != _antiLangmuir.size()) || (_kA.size() < nComp))
		throw InvalidParameterException("MCAL_KA, MCAL_KD, MCAL_QMAX, and MCAL_ANTILANGMUIR have to have the same size");

	return true;
}

inline const char* ExtAntiLangmuirParamHandler::identifier() CADET_NOEXCEPT { return "EXT_MULTI_COMPONENT_ANTILANGMUIR"; }

inline bool ExtAntiLangmuirParamHandler::validateConfig(unsigned int nComp, unsigned int const* nBoundStates)
{
	if ((_kA.size() != _kD.size()) || (_kA.size() != _qMax.size()) || (_kA.size() != _antiLangmuir.size()) || (_kA.size() < nComp))
		throw InvalidParameterException("MCAL_KA, MCAL_KD, MCAL_QMAX, and MCAL_ANTILANGMUIR have to have the same size");

	return true;
}


/**
 * @brief Defines the multi component Anti-Langmuir binding model
 * @details Implements the Anti-Langmuir adsorption model: \f[ \begin{align} 
 *              \frac{\mathrm{d}q_i}{\mathrm{d}t} &= k_{a,i} c_{p,i} q_{\text{max},i} \left( 1 - \sum_j p_j \frac{q_j}{q_{\text{max},j}} \right) - k_{d,i} q_i,
 *          \end{align} \f]
 *          where the factor @f$ p_j \in \{ -1, 1\} @f$ determines the behavior (@c -1 results in Anti-Langmuir, @c +1 in standard Langmuir).
 *          Multiple bound states are not supported. 
 *          Components without bound state (i.e., non-binding components) are supported.
 * @tparam ParamHandler_t Type that can add support for external function dependence
 */
template <class ParamHandler_t>
class AntiLangmuirBindingBase : public ParamHandlerBindingModelBase<ParamHandler_t>
{
public:

	AntiLangmuirBindingBase() { }
	virtual ~AntiLangmuirBindingBase() CADET_NOEXCEPT { }

	static const char* identifier() { return ParamHandler_t::identifier(); }

	CADET_BINDINGMODELBASE_BOILERPLATE
	
protected:
	using ParamHandlerBindingModelBase<ParamHandler_t>::_paramHandler;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_reactionQuasistationarity;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_nComp;
	using ParamHandlerBindingModelBase<ParamHandler_t>::_nBoundStates;

	virtual bool implementsAnalyticJacobian() const CADET_NOEXCEPT { return true; }

	template <typename StateType, typename CpStateType, typename ResidualType, typename ParamType>
	int fluxImpl(double t, unsigned int secIdx, const ColumnPosition& colPos, StateType const* y,
		CpStateType const* yCp, ResidualType* res, LinearBufferAllocator workSpace) const
	{
		typename ParamHandler_t::ParamsHandle const p = _paramHandler.update(t, secIdx, colPos, _nComp, _nBoundStates, workSpace);

		// Protein flux: -k_{a,i} * c_{p,i} * (1 - \sum q_i / q_{max,i}) + k_{d,i} * q_i
		ResidualType qSum = 1.0;
		unsigned int bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			qSum -= static_cast<ParamType>(p->antiLangmuir[i]) * y[bndIdx] / static_cast<ParamType>(p->qMax[i]);

			// Next bound component
			++bndIdx;
		}

		bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			// Residual
			res[bndIdx] = static_cast<ParamType>(p->kD[i]) * y[bndIdx] - static_cast<ParamType>(p->kA[i]) * yCp[i] * static_cast<ParamType>(p->qMax[i]) * qSum;

			// Next bound component
			++bndIdx;
		}

		return 0;
	}

	template <typename RowIterator>
	void jacobianImpl(double t, unsigned int secIdx, const ColumnPosition& colPos, double const* y, double const* yCp, int offsetCp, RowIterator jac, LinearBufferAllocator workSpace) const
	{
		typename ParamHandler_t::ParamsHandle const p = _paramHandler.update(t, secIdx, colPos, _nComp, _nBoundStates, workSpace);

		// Protein flux: -k_{a,i} * c_{p,i} * (1 - \sum q_i / q_{max,i}) + k_{d,i} * q_i
		double qSum = 1.0;
		int bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			qSum -= static_cast<double>(p->antiLangmuir[i]) * y[bndIdx] / static_cast<double>(p->qMax[i]);

			// Next bound component
			++bndIdx;
		}

		bndIdx = 0;
		for (int i = 0; i < _nComp; ++i)
		{
			// Skip components without bound states (bound state index bndIdx is not advanced)
			if (_nBoundStates[i] == 0)
				continue;

			const double ka = static_cast<double>(p->kA[i]);
			const double kd = static_cast<double>(p->kD[i]);

			// dres_i / dc_{p,i}
			jac[i - bndIdx - offsetCp] = -ka * static_cast<double>(p->qMax[i]) * qSum;
			// Getting to c_{p,i}: -bndIdx takes us to q_0, another -offsetCp to c_{p,0} and a +i to c_{p,i}.
			//                     This means jac[i - bndIdx - offsetCp] corresponds to c_{p,i}.

			// Fill dres_i / dq_j
			int bndIdx2 = 0;
			for (int j = 0; j < _nComp; ++j)
			{
				// Skip components without bound states (bound state index bndIdx is not advanced)
				if (_nBoundStates[j] == 0)
					continue;

				// dres_i / dq_j
				jac[bndIdx2 - bndIdx] = ka * yCp[i] * static_cast<double>(p->antiLangmuir[j]) * static_cast<double>(p->qMax[i]) / static_cast<double>(p->qMax[j]);
				// Getting to q_j: -bndIdx takes us to q_0, another +bndIdx2 to q_j. This means jac[bndIdx2 - bndIdx] corresponds to q_j.

				++bndIdx2;
			}

			// Add to dres_i / dq_i
			jac[0] += kd;

			// Advance to next flux and Jacobian row
			++bndIdx;
			++jac;
		}
	}

};

typedef AntiLangmuirBindingBase<AntiLangmuirParamHandler> AntiLangmuirBinding;
typedef AntiLangmuirBindingBase<ExtAntiLangmuirParamHandler> ExternalAntiLangmuirBinding;

namespace binding
{
	void registerAntiLangmuirModel(std::unordered_map<std::string, std::function<model::IBindingModel*()>>& bindings)
	{
		bindings[AntiLangmuirBinding::identifier()] = []() { return new AntiLangmuirBinding(); };
		bindings[ExternalAntiLangmuirBinding::identifier()] = []() { return new ExternalAntiLangmuirBinding(); };
	}
}  // namespace binding

}  // namespace model

}  // namespace cadet

File: src/cadet-mi/bindings/AntiLangmuirBinding.m

classdef AntiLangmuirBinding < KineticQuasiStationaryBindingModel
	%AntiLangmuirBinding Anti-Langmuir binding model
	%
	% See also BINDINGMODEL

	% Copyright: (C) 2008-2024 The CADET Authors
	%            See the license note at the end of the file.

	properties(Constant)
		name = 'MULTI_COMPONENT_ANTILANGMUIR'; % Name of the binding model according to CADET file format specs
	end

	properties (Dependent, Transient)
		% Adsorption rate in [m^3_MP / (mol * s)]
		kA;
		MCAL_KA;
		% Desorption rate in [1 / s]
		kD;
		MCAL_KD;
		% Capacity in [mol / m^3_SP]
		qMax;
		MCAL_QMAX;
		% Anti-Langmuir coefficient [-]
		antiFactor;
		MCAL_ANTILANGMUIR;
	end

	methods

		function obj = AntiLangmuirBinding(kA, kD, qMax, antiFactor)
			%ANTILANGMUIRBINDING Constructs an Anti-Langmuir binding model object
			%   OBJ = ANTILANGMUIRBINDING(KA) creates an Anti-Langmuir binding model with
			%   the given adsorption rates KA.
			%
			%   OBJ = ANTILANGMUIRBINDING(..., KD) also sets the desorption rates to KD.
			%
			%   OBJ = ANTILANGMUIRBINDING(..., KD, QMAX) also sets the desorption rates to KD
			%   and the capacity to QMAX.
			%
			%   OBJ = ANTILANGMUIRBINDING(..., KD, QMAX, ANTIFACTOR) also sets the desorption
			%   rates to KD, the capacity to QMAX, and the anti-Langmuir coefficients to ANTIFACTOR.

			obj = obj@KineticQuasiStationaryBindingModel();

			if nargin >= 1
				obj.kA = kA;
			end
			if nargin >= 2
				obj.kD = kD;
			end
			if nargin >= 3
				obj.qMax = qMax;
			end
			if nargin >= 4
				obj.antiFactor = antiFactor;
			end
		end

		function res = validate(obj, nComponents, nBoundStates)
			%VALIDATE Validates the binding model parameters
			%   RES = VALIDATE(NCOMPONENTS, NBOUNDSTATES) uses the number of
			%   components NCOMPONENTS of the model and the number of bound
			%   states of each of these components as given by the vector
			%   NBOUNDSTATES to validate the parameters of the binding model.
			%   Returns true in RES if everything is fine and false otherwise.

			validateattributes(obj.kA, {'double'}, {'vector', 'nonempty', '>=', 0.0, 'finite', 'real', 'numel', nComponents}, '', 'kA');
			validateattributes(obj.kD, {'double'}, {'vector', 'nonempty', '>=', 0.0, 'finite', 'real', 'numel', nComponents}, '', 'kD');
			validateattributes(obj.qMax, {'double'}, {'vector', 'nonempty', '>', 0.0, 'finite', 'real', 'numel', nComponents}, '', 'qMax');
			validateattributes(obj.antiFactor, {'double'}, {'vector', 'nonempty', '>=', -1.0, '<=', 1.0, 'finite', 'real', 'numel', nComponents}, '', 'antiFactor');
			if ~all((obj.antiFactor == -1) | (obj.antiFactor == 1))
				error('CADET:invalidConfig', 'Expected antiFactor to be an array consisting of -1 and 1.')
			end
			res = obj.validate@KineticQuasiStationaryBindingModel(nComponents, nBoundStates);
		end

		function val = get.kA(obj)
			val = obj.data.MCAL_KA;
		end

		function set.kA(obj, val)
			validateattributes(val, {'double'}, {'vector', 'nonempty', '>=', 0.0, 'finite', 'real'}, '', 'kA');
			obj.data.MCAL_KA = val;
			obj.hasChanged = true;
		end

		function val = get.kD(obj)
			val = obj.data.MCAL_KD;
		end

		function set.kD(obj, val)
			validateattributes(val, {'double'}, {'vector', 'nonempty', '>=', 0.0, 'finite', 'real'}, '', 'kD');
			obj.data.MCAL_KD = val;
			obj.hasChanged = true;
		end

		function val = get.qMax(obj)
			val = obj.data.MCAL_QMAX;
		end

		function set.qMax(obj, val)
			validateattributes(val, {'double'}, {'vector', 'nonempty', '>', 0.0, 'finite', 'real'}, '', 'qMax');
			obj.data.MCAL_QMAX = val;
			obj.hasChanged = true;
		end

		function val = get.antiFactor(obj)
			val = obj.data.MCAL_ANTILANGMUIR;
		end

		function set.antiFactor(obj, val)
			validateattributes(val, {'double'}, {'vector', 'nonempty', 'finite', 'real'}, '', 'antiFactor');
			if ~all((val == -1) | (val == 1))
				error('CADET:invalidConfig', 'Expected antiFactor to be an array consisting of -1 and 1.')
			end
			obj.data.MCAL_ANTILANGMUIR = val;
			obj.hasChanged = true;
		end

		function val = get.MCAL_KA(obj)
			val = obj.kA;
		end
		function set.MCAL_KA(obj, val)
			obj.kA = val;
		end
		function val = get.MCAL_KD(obj)
			val = obj.kD;
		end
		function set.MCAL_KD(obj, val)
			obj.kD = val;
		end
		function val = get.MCAL_QMAX(obj)
			val = obj.qMax;
		end
		function set.MCAL_QMAX(obj, val)
			obj.qMax = val;
		end
		function val = get.MCAL_ANTILANGMUIR(obj)
			val = obj.antiFactor;
		end
		function set.MCAL_ANTILANGMUIR(obj, val)
			obj.antiFactor = val;
		end

	end

	methods (Static)

		function obj = loadobj(S)
			if isstruct(S)
				obj = AntiLangmuirBinding();
				obj.loadobjInternal(S);
			end
		end
		
	end

end

% =============================================================================
%  CADET
%  
%  Copyright (C) 2008-2024: The CADET Authors
%            Please see the AUTHORS and CONTRIBUTORS file.
%  
%  All rights reserved. obj program and the accompanying materials
%  are made available under the terms of the GNU Public License v3.0 (or, at
%  your option, any later version) which accompanies obj distribution, and
%  is available at http://www.gnu.org/licenses/gpl.html
% =============================================================================
